internal/allocator/allocator.go

```diff
@@ -381,3 +372,168 @@ func ipConfusesBuggyFirmwares(ip net.IP) bool {
 	}
 	return ip[3] == 0 || ip[3] == 255
 }
+
+// Dual-stack.
+
+// AssignDual assigns the requested ip's to svc, if the assignment is
+// permissible by sharingKey and backendKey.
+func (a *Allocator) AssignDual(svc string, ip, ip2 net.IP, ports []Port, sharingKey, backendKey string) error {
+	if ipIsIPv6(ip) == ipIsIPv6(ip2) {
+		return fmt.Errorf("%q %q is has the same family", ip, ip2)
+	}
+	pool := poolFor(a.pools, ip)
+	if pool == "" {
+		return fmt.Errorf("%q is not allowed in config", ip)
+	}
+	if poolFor(a.pools, ip2) != pool {
+		return fmt.Errorf("%q %q different pools", ip, ip2)
+	}
+
+	sk := &key{
+		sharing: sharingKey,
+		backend: backendKey,
+	}
+
+	if err := a.checkSharing(svc, ip.String(), ports, sk); err != nil {
+		return err
+	}
+	if err := a.checkSharing(svc, ip2.String(), ports, sk); err != nil {
+		return err
+	}
+
+	// Either the IP is entirely unused, or the requested use is
+	// compatible with existing uses. Assign! But unassign first, in
+	// case we're mutating an existing service (see the "already have
+	// an allocation" block above). Unassigning is idempotent, so it's
+	// unconditionally safe to do.
+	alloc := &alloc{
+		pool:  pool,
+		ip:    ip,
+		ip2:   ip2,
+		ports: make([]Port, len(ports)),
+		key:   *sk,
+	}
+	for i, port := range ports {
+		port := port
+		alloc.ports[i] = port
+	}
+	a.assign(svc, alloc)
+	return nil
+}
+
+// AllocateFromPoolDual assigns available IP's from pool to service.
+func (a *Allocator) AllocateFromPoolDual(svc string, poolName string, ports []Port, sharingKey, backendKey string) (net.IP, net.IP, error) {
+	if alloc := a.allocated[svc]; alloc != nil {
+		if err := a.AssignDual(svc, alloc.ip, alloc.ip2, ports, sharingKey, backendKey); err != nil {
+			return nil, nil, err
+		}
+		return alloc.ip, alloc.ip2, nil
+	}
+
+	fmt.Println("AllocateFromPoolDual", svc, poolName)
+	pool := a.pools[poolName]
+	if pool == nil {
+		return nil, nil, fmt.Errorf("unknown pool %q", poolName)
+	}
+
+	// Find IP's from both families
+	var ip, ip2 net.IP
+	sk := &key{
+		sharing: sharingKey,
+		backend: backendKey,
+	}
+	for _, cidr := range pool.CIDR {
+		if ip == nil {
+			c := ipaddr.NewCursor([]ipaddr.Prefix{*ipaddr.NewPrefix(cidr)})
+			for pos := c.First(); pos != nil; pos = c.Next() {
+				if pool.AvoidBuggyIPs && ipConfusesBuggyFirmwares(pos.IP) {
+					continue
+				}
+				if a.checkSharing(svc, pos.IP.String(), ports, sk) != nil {
+					continue
+				}
+				ip = pos.IP
+				break
+			}
+		} else {
+			if cidrIsIPv6(cidr) == ipIsIPv6(ip) {
+				continue
```

Comment:
is there a special order assumed here I am bit confused with v6 check here and skip the allocation ?

Reply:
No, any order will do, ipv4,ipv6 or ipv6,ipv4. The `cidrIsIPv6(cidr)` == ipIsIPv6(ip)` ensures that we pick ip2 from a cidr of the "other" family, whatever that may be.